Meadowlark Interface for PCIe

darikoneil opened the pull request:
**I have modified the `Meadowlark` class to be compatible with both HDMI & PCIe models in response to issue #88 .** 

This PR is currently a draft to get feedback before testing it on physical SLMs. The file was linted, with Ruff (see ruleset at end of PR--pylint was hurting my eyes), formatted with black, and it has no breaking changes to the API. 

**Summary of Changes**

- The `sdk_path` argument now defaults to `None`. Some of the SDKs install in slightly different locations, so I added a method to search through the `C:\Program Files\Meadowlark Optics\` for the SDK. If multiple `Blink_C_wrapper.dll` files are found, it will warn the user and default to the most recently created file.
- There is now an `slm_number` argument added to the constructor and various methods to support having multiple SLMs on the same machine. This supported only for the PCIe models. The number of boards available is stored as a private attribute. Oddly, I don't see a method to collect this after initializing the SDK. Therefore, I had to store the value returned during initialization for use with the `info` method (which is no longer static)
- To more gracefully handle having (at least) three different SDKs, the constructor will attempt to initialize each SDK option until it succeeds or runs out of options. Each initialization is in a separate private method to keep the logic clean--especially if it turns out that even more are needed. I made a private enum `SDK_MODE` that stores which SDK was initialized. The various methods can then pattern match to call the appropriate method.
- There's actually two more SDKs that I did not implement. (1) You can initialize with a GPU to use the Gerchberg-Saxton algo, but those are implemented in `slmsuite`. So there's not a good reason (imo) to offer that option to the user and add complexity to the constructor.  (2) I did not implement the SDK using the GPU (Overdrive Plus). If the ultra-high speed boards are using it, it can be implemented. However, according to my SDK manuals it's only offered on the legacy 512x512 models, it requires some annoying complexity to be dealt with (loading LUT only at time of initialization, some additional parameters & methods), and anyone needing that much speed is probably pre-loading their images to load with an external trigger. So, I didn't include that if it most likely will not be touched, but would still need tested & maintained.
- One other thing to note on code-style: the meadowlark file now contains type-hints. I can remove them if you want. I figured I'd give you the option of whether to leave them in, since it makes static analysis tools / IDEs / autocomplete a bit more helpful.

**Ruff Configuration**
[lint]
select = [
    # pycodestyle
    "E",
    # Pyflakes
    "F",
    "W",
    # pyupgrade
    "UP",
    # flake8-bugbear
    "B",
    #flake8 builtins
    "A",
    # flake blind expect
    "BLE",
    # flake8-simplify
    "SIM",
    # flake8-annotations
    "ANN",
    # pylint error
    "PLE",
    #pylint warnings
    "PLW",
]
ignore =[
    "ANN003",
    "ANN204"
]

ichristen commented:
Awesome! I'm taking a look now!

First small comment: does the same Blink SDK .dll support both HDMI and PCIe SLMs? Currently, you're looking for the most recent .dll, but this might cause issues if a user has both HDMI and PCIe SLMs connected. The system would need to load the .dll appropriate for HDMI or PCIe?

ichristen commented:
I added a few comments as part of the commit labeled with "TODO".

I'm good with pushing this to main once you've tested the code on your SLMs.

Thanks much for your good work on this!

ichristen commented:
Also, this provides impetus to add type hints to the rest of slmsuite. Probably best to move from sphinx type documentation to some autodoc'ed version of the python type hints e.g. https://github.com/tox-dev/sphinx-autodoc-typehints.

darikoneil commented:
> Also, this provides impetus to add type hints to the rest of slmsuite. Probably best to move from sphinx type documentation to some autodoc'ed version of the python type hints e.g. https://github.com/tox-dev/sphinx-autodoc-typehints.

Since you already have types in the documentation, it should be extremely easy. I think GitHub copilot could type-hint the whole thing accurately (obviously one would still want to double check it all). I could test that if you want since it would only take a few seconds.

darikoneil commented:
Hi, 

Sorry about the delay in testing this--I had some unexpected chaos w.r.t. to my research funding. I addressed the TODO's where appropriate with #Review and added a #Warn explaining the 'flip immediate' option. I made a few minor corrections that arose during testing. There is one big change (and the reason I haven't marked this ready to merge). I discovered that an unsuccessful attempt to use the 'Create_SDK' method can cause subsequent attempts to fail (even if the provided values ought to be successful).

I can think of two options.:
1. We can to try to unload the DLL. Unfortunately, it seems this is isn't exactly straightforward. I added a method to unload the DLL after each attempt and some related refactoring. In my testing, I am able to recover from unsuccessful attempts on Windows. I don't have the ability to test an SLM on a Linux machine. I'm also not sure if an HDMI SLM will load successfully after having attempted to load the PCIe SLMs.  Are you able to test if this works with an HDMI SLM? 
2. We can attempt to load the SDKs in their own subprocesses when attempting to identify the correct one. When the correct one is identified, we can load the SLM in the main process. This should ensure that each loading is a completely new instance. I'm 95% sure this will work, but I haven't tried this option yet.

ichristen commented:
Yo, @darikoneil , no worries. Sorry to hear about funding chaos; I hope it was fully resolved.

Regarding the .dll load, what do you think about an option 3:

3. Reinstate a `sdk_mode=None` argument to the class constructor. The default `None` would parse the `sdk_path` to determine whether the installation is a HDMI/PCI/etc .dll and load appropriately (do you think this is possible?). If it can't tell, then the constructor would error, requesting that the user specify `sdk_mode="HDMI"`/`sdk_mode="PCIe"`/etc. Along the same lines, I think if `sdk_path=None` and `_locate_blink_c_wrapper` returns multiple paths, the user should also be asked to specify `sdk_path` from options returned by the error message.

nobias commented:
Hi everyone! Thanks for your work on this, this has been very helpful for us. Let me jump in and say that we also ran into the problem that earlier failed attempts to call Create_SDK() lead to subsequent calls failing always. 
What's more, we realized that there is yet another version of the SDK, newer than the "pcie_modern" one that you implemented. This newest version (let's call it "pcie_postmodern" for now) shipped with the USB drive included with our latest SLM, a 1024x1024 "high speed" model with PCIe interface. The API difference ist that Create_SDK now only takes two arguments, and Write_image() is also different. The signatures are:

		void Create_SDK (unsigned int* n_boards_found, int *constructed_ok);

		int Write_image (int board, unsigned char* image, int wait_for_trigger, int
		flip_immediate, int output_pulse_image_flip, unsigned int trigger_timeout_ms);

I can provide screenshots of the API docs tomorrow. Would you like me to try and add our code for the "postmodern_pcie" API into this pull request?

I would support adding a constructor argument for selecting the SDK version.

ichristen commented:
Hi @nobias ,

I think the [version of Write_image](https://github.com/slmsuite/slmsuite/blob/afc4197e2a9dcbc12a4c0faf2e3a88f9ce32b5dd/slmsuite/hardware/slms/meadowlark.py#L681) that @darikoneil implemented conforms with the signature?

Regardless, this versioning is beginning to become too complicated for a GitHub thread. I'm E-mailing both of you (along with Meadowlark) with a link to a Google Sheet containing a table of each SDK version alongside what is supported. When we're happy with the table, we can export it as .csv and [include it in the docs](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/tables.html#csv-files). Could you link the .pdf of the API documentation that Meadowlark sent you in that Sheet?

I realized that I forgot to respond to @darikoneil 's question of whether I can test with an HDMI SLM. Yes, I can test and I can also upload the docs that the SLM came with, I think.

nobias commented:
The signatures of Write_image() are very similar, but in my latest "pcie_postmodern" version, it has one argument less: It doesn't expect unsigned int image_size as the third argument anymore.

ichristen commented:
Gotcha, and thank you, I missed that.

ichristen commented:
This is a very hacky solution, but we could use [`pefile`](https://github.com/erocarrera/pefile) to examine each .dll pre-import. Unfortunately, `pefile` cannot read the number of arguments and the argument types for exported functions. It seems from a brief search that there is no reliable standard to extract this information from .dlls. What `pefile` does provide, however, is the name of each function and the address (in the .dll binary) at which each function is defined.

Thus, we could maybe analyze Meadowlark .dlls for a "signature" mapping function names to addresses and keep a library of recognized signatures corresponding to the common HDMI dll, PCIe dll, etc, in the case of `sdk_mode=None`. If a signature is not recognized, the user is prompted to select an `sdk_mode`.

ichristen commented:
A slightly less hacky solution might be to use a .dll's hash code. This would involve a mapping of hash -> `sdk_mode` saved in meadowlark.py. If a user tries to load an unrecognized hash with `sdk_mode=None`, the error message would encourage them to pull request to add their hash code to the mapping. This implementation and the one above depends on how many different .dll versions Meadowlark commonly distributed over the years. If only O(10) different types were commonly distributed, then maybe it's reasonable. If it's O(100), then less so.

darikoneil commented:
@ichristen @nobias yes, Meadowlark could be helpful her, agree on the email.

@ichristen I think parsing the header file is an excellent idea, if that's what you're proposing. If you're not, then that is what I am proposing.

I have a couple 'old' SDKs, and it looks like the parameters themselves have actually stayed pretty consistent in their typing and naming? Further, many of them are configuration options. Rather than having to keep a map of known signatures, we might be able to dump all the known configuration options into a data structure, dynamically parse the SDK to get the signatures, and bind the 'static' configurations to many of these functions such that they are all essentially identical.

For example:
```
  int Write_image(int board,
                  unsigned char* image,
                  unsigned int image_size,
                  int wait_for_trigger,
                  int flip_immediate,
                  int output_pulse_image_flip,
                  int output_pulse_image_refresh,
                  unsigned int trigger_timeout_ms);
```

The write image obviously always requires the image. The rest of the parameters are either explicitly static for a single instance (like the board number or image size), or configurations that are not currently exposed to `slmsuite` users. On instantiation, you can just bind whichever additional parameters are required and then have the public facing method call the partial function.

ichristen commented:
@darikoneil Is the header file for the .dll always included in the installation? I don't have a installation in front of me and I don't remember.

nobias commented:
I have access to two different SDK installations, and in both cases, there is a Blink_C_wrapper.h (and also a Blink_SDK.h right next to the respective .dll files. In my case, they are all in `C:\Program Files\Meadowlark Optics\Blink Plus\SDK`.
One of these installs is of the "pcie_modern" variety (Create_SDK(): 8 arguments, Write_image(): 8), the other one is the even more recent version that I reported in the pull request thread (Create_SDK(): 2 arguments, Write_image(): 6). So at least in these relatively recent SDK versions, parsing the .h files sounds like a good idea to me!

darikoneil commented:
@ichristen @nobias I have several SDK versions and they all include header files.

ichristen commented:
@darikoneil , @nobias , I revised the class to parse the headers to determine what SDK mode a given .dll/.h is compatible with. I tested header parsing on the three SDK cases (conforming with Meadowlark's suggestion), but I did not test on actual hardware yet.

A main difference is that SDKs are now stored as singleton *class attributes*, one for each type of SDK mode and shared among all possible connected SLMs.

Let me know what you think!

darikoneil commented:
> @darikoneil , @nobias , I revised the class to parse the headers to determine what SDK mode a given .dll/.h is compatible with. I tested header parsing on the three SDK cases (conforming with Meadowlark's suggestion), but I did not test on actual hardware yet.
> 
> A main difference is that SDKs are now stored as singleton _class attributes_, one for each type of SDK mode and shared among all possible connected SLMs.
> 
> Let me know what you think!

Sorry for being MIA, I'm trying to ~~cram~~ write my thesis at the moment. I just took a look and it looks great! I haven't tested it yet; I believe need to get the latest SDK from meadowlark first. Sending an email later today regarding holoeye.

ichristen commented:
Hi @nobias , @darikoneil , 

There are also `Hamamatsu` (#100) and `Holoeye` (#105) pull requests in the pipeline. I was wondering if you might have time to test to quickly test the new `Meadowlark` class on PCIe hardware. Thinking about trying to pull everything together into a SLM-hardware release over the weekend.

Best,
Ian

darikoneil commented:
> Hi @nobias , @darikoneil ,
> 
> There are also `Hamamatsu` (#100) and `Holoeye` (#105) pull requests in the pipeline. I was wondering if you might have time to test to quickly test the new `Meadowlark` class on PCIe hardware. Thinking about trying to pull everything together into a SLM-hardware release over the weekend.
> 
> Best, Ian

Hi @ichristen,

I've penciled in a test for Saturday morning on our 1920x1152; I can test the 512x512 today. 

D

nobias commented:
Hi,
I would love to contribute, but I'm afraid I won't be able to test in time for this weekend. We're in the process of incorporating our Meadowlark into a much larger setup, and it is a bit too difficult right now to wire it up and test. Apologies. I think I'd get to it early-mid next week. I'll let you know as soon as I have some results. Sorry!
Best,
Tobias

> On Apr 25, 2025, at 07:29, Darik A O’Neil ***@***.***> wrote:
> 
> Caution: External email darikoneil left a comment (slmsuite/slmsuite#90)
> Hi @nobias , @darikoneil ,
> There are also Hamamatsu (#100) and Holoeye (#105) pull requests in the pipeline. I was wondering if you might have time to test to quickly test the new Meadowlark class on PCIe hardware. Thinking about trying to pull everything together into a SLM-hardware release over the weekend.
> Best, Ian
> Hi @ichristen,
> I've penciled in a test for Saturday morning on our 1920x1152; I can test the 512x512 today.
> D
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>

darikoneil commented:
I ran the test and fixed some bugs so that appeared. It seems to work for my legacy SLM!

The newer 1920x1152 was being registered as legacy; I think there might be some shenanigans on that SLM's PC, where an older SDK was pasted into another one. I will ask meadowlark for the most recent SDK ASAP. It might be worth noting in the documentation that the parser will fail without raising an exception if someone is using the legacy SDK for the larger, newer models.

ichristen commented:
@darikoneil I will send you the suggested SDKs over E-mail / Google Drive in a few minutes. Thanks for testing this!

ichristen commented:
Actually, @darikoneil , I think that the 1920 x 1152 being registered as legacy is correct per the table (see .csv and Google Sheet). If I recall correctly, Meadowlark no longer supports this model in the suggested non-legacy PCIe Blink version (1.1.4.870). They only support the most modern 1024 x 1024 version on PCIe. It would be interesting to get verification of this.

darikoneil commented:
@ichristen 
 
 Excellent, I got your email and see your TODO on OverDrive Plus. I think there might be a need to have a separate sdk_mode for OverDrivePlus & Legacy, as it seems the overdrive plus model has a different initialization process even when not actually using the OverDrive Plus features.

Alternatively, we could have a check that runs the alternative initialization if the sdk_mode is legacy and the SLM shape is exactly 512x512. 

A short-term possibility to still get meadowlark in this update is we can temporarily drop support for 512x512. I can adjust the legacy SDK to behave as expected for the non-OverDrivePlus 512x512 SLMs, have an exception raised if sdk_mode is legacy & the SLM shape is 512x512, and I’ll test to confirm it now works on the 1920x1152.

We can have another update re-add 512x512 support and do any refactoring that might need to be done to accommodate that.

nobias commented:
Hi! I just tested with my HSPDM1K-950-1300-PC8 and Blink SDK 1.1.4.870. Your current code correctly identified the SDK as PCIE_MODERN and loaded the DLL. However, if I didn't confuse myself entirely, it seems that in this case you call Create_SDK() with 8 arguments (line 678). Err, so, isn't this supposed to be the mode where this function has two arguments? I can push a commit to this PR that implements this call with 2 args and that works with my SDK. I just wanted to make sure that that's what you actually had intended to do here?

nobias commented:
Here's a patch for this change. Also contains two other fixes regarding checking return/output arguments. [meadowlark_2args.patch](https://github.com/user-attachments/files/20005813/meadowlark_2args.patch)

ichristen commented:
@nobias It looks like I didn't implement the case to switch depending on `Meadowlark._slm_lib_trace[self.sdk_mode][0]` for Create_SDK. There should be both three and six argument cases for modern. I added more documentation in a commit just now (I realized `_SLM_LIB_TRACES` was not documented).

nobias commented:
Ah, fantastic. The latest commit works perfectly for us. I tested loading LUTs and writing phase patterns, all seems to work great for HSPDM1K-950-1300-PC8 and Blink SDK 1.1.4.870. Thank you so much! Please let me know if you'd like me to run more specific tests.

ichristen commented:
Nice work everyone! Merging now.
